Add a slurm workflow manager

linsword13 opened the pull request:
Example usage:
    
```yaml
ramble:
  variants:
    workflow_manager: slurm
  variables:
    processes_per_node: 1
  applications:
    hostname:
      workloads:
        local:
          experiments:
            test:
              variables:
                n_nodes: 1
```

```
  # The batch_submit is defined by the workflow object
ramble on

  # Executors for query and cancel are available
ramble on --executor "{batch_query}"
ramble on --executor "{batch_cancel}"

  # The status report is also available in analyze
ramble workspace analyze -p
```

douglasjacobsen commented on var/ramble/repos/builtin/workflow_managers/slurm/workflow_manager.py:
```diff
@@ -0,0 +1,108 @@
+    tags("workflow", "slurm")
+
+    render_content(
+        name="batch_submit",
```
How does this work when we define `batch_submit` in the variables? Does this override the definition of the `batch_submit` variable?

linsword13 replied:
Yeah it does override the `batch_submit` variable. I do that so that `ramble on` works without needing to specify an `executor`. Do you see this being undesirable in some cases?

douglasjacobsen replied:
Nope, I was more concerned that the variable would override the template.

douglasjacobsen commented on var/ramble/repos/builtin/workflow_managers/slurm/workflow_manager.py:
```diff
@@ -0,0 +1,197 @@
+        name="slurm_execute_experiment", content_func="_slurm_execute_script"
+    )
+
+    def _slurm_execute_script(self):
```
Can we maybe move this into the base class, as a general "Build pragmas" function, just since I think this will be used fairly frequently.

Maybe it should be something other than pragmas specifically though... more like something around conditional expansion?

linsword13 replied:
Makes sense, updated.

douglasjacobsen commented on var/ramble/repos/builtin/workflow_managers/slurm/workflow_manager.py:
```diff
@@ -0,0 +1,197 @@
+
+    def generate_query_command(self, job_id):
+        return rf"""
+status=$(squeue -h -o "%t" -j {job_id})
```
I'm not sure how picky I am about this, but I would almost prefer the status map be in python rather than in bash.

And maybe some of the logic mapped into the base class too.

linsword13 replied:
Good call. I moved the get_status call inside the runner to be in python instead of relying on this bash script, so now `ramble workspace analyze` would rely on python. Do you prefer even the `--executor "{batch_query}"` to be handled in python instead of the bash script right now?

douglasjacobsen commented on lib/ramble/ramble/test/conftest.py:
```diff
@@ -187,18 +199,25 @@ def mock_modifiers(mock_mods_repo_path):
+
+
+@pytest.fixture(scope="function")
+def mock_workflow_managers(mock_wms_repo_path):
```
Not related to this PR, mostly just something to think about.

Do you have any ideas on how we can clean up this stuff, and make it so we don't have to define new fixtures for each object type?

linsword13 replied:
We can perhaps have a fixture that defaults to all the builtin paths, and can be overridden to point to builtin.mock path for specific object type. But that makes the test itself slightly more verbose, so not sure if that's much of an improvement.

douglasjacobsen replied:
OK cool, thanks. Maybe we can keep this in the back of our minds to see if we can make this nicer in the future.

douglasjacobsen commented on var/ramble/repos/builtin/workflow_managers/slurm/workflow_manager.py:
```diff
@@ -0,0 +1,197 @@
+        default="",
+        description="Extra sbatch headers added to slurm job script",
+    )
+
```
Can you add a `workflow_manager_variable` to define `hostlist` as `$SLURM_JOB_NODELIST` too? That might make composition with modifiers that require this to be define a bit easier.

linsword13 replied:
Great suggestion, added.

douglasjacobsen commented on var/ramble/repos/builtin/workflow_managers/slurm/workflow_manager.py:
```diff
@@ -0,0 +1,197 @@
+            ("#SBATCH -N {}", "n_nodes"),
+            ("#SBATCH -p {}", "partition"),
+            ("#SBATCH --ntasks-per-node {}", "processes_per_node"),
+            ("#SBATCH -J {}", "job_name"),
```
Can you also add `--gpus-per-node` = `{gpus_per_node}` here?

linsword13 replied:
Definitely, added.

douglasjacobsen commented on var/ramble/repos/builtin/workflow_managers/slurm/workflow_manager.py:
```diff
@@ -0,0 +1,197 @@
+        super().__init__(file_path)
+
+        self.runner = SlurmRunner()
+
```
We should also think about how to allow this modifier to have a default `mpi_command` (that could still be overridden)

linsword13 replied:
Added. It works to add that as a `workflow_variable`, and that's overridable.

linsword13 commented:
Marking this as a draft, given that this PR will need to be updated if https://github.com/GoogleCloudPlatform/ramble/pull/804 is in.

douglasjacobsen commented:
Something I just found out, if a job is completed and you run `ramble on --executor='{query_job}'` then it'll say the job failed (even if you haven't analyzed it.

We might want to see if there's a way to use something like scontrol to get the status of completed jobs.

linsword13 commented:
> Something I just found out, if a job is completed and you run `ramble on --executor='{query_job}'` then it'll say the job failed (even if you haven't analyzed it.
> 
> We might want to see if there's a way to use something like scontrol to get the status of completed jobs.

Hm that's unexpected. I did have a `sacct` command to try to query for status of completed jobs, if `squeue` comes up empty. I will look into it.

douglasjacobsen commented:
Let me try it again, it might be that analyzing the job failed for some reason.

douglasjacobsen commented:
Yeah, I think I just had a weird issue with my experiments.

Can we also add default `-o` and `-e` pragmas to the generated submission script? This can place the slurm logs into the experiment directory.

linsword13 committed:
Scaffolding for workflow_manager object type

linsword13 committed:
Implement slurm workflow manager

linsword13 committed:
Include in workflow status in workspace analyze

linsword13 committed:
Add tests

linsword13 committed:
Cleanup a unused variable

linsword13 commented:
> Yeah, I think I just had a weird issue with my experiments.
> 
> Can we also add default `-o` and `-e` pragmas to the generated submission script? This can place the slurm logs into the experiment directory.

Thanks for trying out! Also added the `-o` and `-e`.

linsword13 commented:
> Marking this as a draft, given that this PR will need to be updated if #804 is in.

This is now updated. It currently uses changes in https://github.com/GoogleCloudPlatform/ramble/pull/814.

linsword13 committed:
Add var def for `mpi_command`, serving as a default

linsword13 committed:
Update completion

douglasjacobsen commented on lib/ramble/ramble/application.py:
```diff
@@ -65,7 +65,8 @@
 experiment_status = Enum(
-    "experiment_status", ["UNKNOWN", "SETUP", "RUNNING", "COMPLETE", "SUCCESS", "FAILED"]
+    "experiment_status",
+    ["UNKNOWN", "SETUP", "RUNNING", "COMPLETE", "SUCCESS", "FAILED", "CANCELLED"],
```
Can we add a "SUBMITTED" status between "SETUP" and "RUNNING?

linsword13 replied:
Definitely, added.

douglasjacobsen reviewed (changes requested):
This is looking great. One small request. :)

linsword13 committed:
Add a `SUBMITTED` ramble status

Also generate the bash associative array declaration from the python
status_map.
